formatting for button columns, join support, form show_id

jpsteil opened the pull request:
Suggested fixes for a few issues I ran into:

1. The button column on the right needs special formatting for it to remain narrow, only taking the space it needs for the buttons. With bulma this is handled with the is-narrow class. To accommodate, I subclassed Column into ButtonsColumn so I could apply different formatting to the <td> for that column.  Not sure this is the implementation you want.
2. _make_action_buttons didn't work with joins because row.id doesn't know about joins and the self.param.field_id. I updated so it will use self.param.field_id if it contains a value, if not it defaults to row.id. In addition, self.param.field_id was not specified unless the user defined on instantiation. I updated the process() method to use the ID of the first table found if no field_id was supplied. QUESTION - is this the right behavior, or should we raise an error if left is specified without field_id?
3. show_id is being passed to the form.  The code to enforce this in form.py was indented one level where it shouldn't have been.

One more issue for discussion is how we handle pre/post action buttons.  It looks to me like if we specify a button as a Column that it will appear in its own column.  I feel it should go before or after (depending on how the dev wants it) the standard grid buttons in the same column.  Thoughts?  If you want, I can work on the implementation.

jpsteil commented:
Added another commit that adds column sorting capabilities - feel free to refactor

mdipierro commented:
I agree with 2 and 3. Also if you want to reinstate pro and post I have no objection.

I definitively do not like ButtonColumn. What I would suggest instead is a mechanism to pass to the Column class the HTML class and style of the outer TD element.

jpsteil commented:
I will look at re-implementing the formatting and get rid of ButtonsColumn.

Will also look at reinstating pre/post action buttons.

jpsteil commented:
Refactored earlier change that introduced ButtonsColumn.  ButtonsColumn is now gone.  Instead you can pass the grid_class_style key to be used for the outer TD element.

Added pre/post action buttons back in.

Went through GridClassStyle and GridClassStyleBulma to make sure each one had all the same keys.

Of course - please refactor as needed or let me know if you'd like something implemented differently.

jpsteil commented:
Allow left join fields to co-exist in a grid with Column() specified.

My implementation changes how self.needed_fields is set.  Now, after looking through Columns/FieldVirtual it goes through all fields passed in and makes sure they are included.

jpsteil commented:
I sometimes use grids that have create, details, editable, deletable all False.  In this case, we still want to display the grid.  Escpecially now with Column(), users may want to provide all their own buttons and not use any of ours.

mdipierro commented:
LGTM. This depends on #593 and I will merge as soon as the other one is approved. Or do you prefer we merge into the branch and make it a single PR?

jpsteil commented:
I was thinking to merge this in to the grid-improvements branch.  But, I'm not that well-versed in git so whichever is easier is fine with me.

mdipierro commented:
I can do it now. This week i have limited connectivity but overall more
time to work on py4web. :-)


On Wed, Jul 14, 2021, 20:47 Jim Steil ***@***.***> wrote:

> I was thinking to merge this in to the grid-improvements branch. But, I'm
> not that well-versed in git so whichever is easier is fine with me.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/web2py/py4web/pull/596#issuecomment-880127705>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHLZTZUH6O4DRWC6ZTWVALTXXLVJANCNFSM5AJNFR3Q>
> .
>

mdipierro commented:
Actually you already made the pr against grid-improvements. You are way
ahead of me. Thanks. Merged

On Wed, Jul 14, 2021, 20:55 Massimo DiPierro ***@***.***>
wrote:

> I can do it now. This week i have limited connectivity but overall more
> time to work on py4web. :-)
>
>
> On Wed, Jul 14, 2021, 20:47 Jim Steil ***@***.***> wrote:
>
>> I was thinking to merge this in to the grid-improvements branch. But, I'm
>> not that well-versed in git so whichever is easier is fine with me.
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/web2py/py4web/pull/596#issuecomment-880127705>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAHLZTZUH6O4DRWC6ZTWVALTXXLVJANCNFSM5AJNFR3Q>
>> .
>>
>
